lib/services/reconciler.go

```diff
@@ -53,6 +53,10 @@ type ReconcilerConfig[T Reconciled] struct {
 // Matcher is used by reconciler to match resources.
 type Matcher[T any] func(T) bool
 
+// Updater is a function type used by a reconciler to update resources to
+// a new value
+type Updater[T Reconciled] func(ctx context.Context, new, old T) error
```

Comment:
nit:
Does this function type  need to be exported ?

Also I would add a comment that old value can be `nil`